libs/consumption_metrics/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+//!
+//! Shared code for consumption metics collection
+//!
+
+use utils::id::TimelineId;
+
+use utils::id::TenantId;
+
+use serde::Serialize;
+
+use chrono::{DateTime, Utc};
+use rand::Rng;
+
+#[derive(Serialize, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+#[serde(tag = "type")]
+pub enum EventType {
+    #[serde(rename = "absolute")]
+    Absolute { time: DateTime<Utc> },
+    #[serde(rename = "incremental")]
+    Incremental {
+        start_time: DateTime<Utc>,
+        stop_time: DateTime<Utc>,
+    },
+}
+
+#[derive(Serialize, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Event<Extra> {
+    #[serde(flatten)]
+    #[serde(rename = "type")]
+    pub kind: EventType,
+
+    pub metric: &'static str,
+    pub idempotency_key: String,
+    pub value: u64,
+
+    #[serde(flatten)]
+    pub extra: Extra,
+}
+
+/// PageserverConsumptionMetric struct that defines the format for one
+/// storage metric entry
+/// i.e.
+///
+/// ```json
+/// {
+/// "metric": "remote_storage_size",
+/// "type": "absolute",
+/// "tenant_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+/// "timeline_id": "a03ebb4f5922a1c56ff7485cc8854143",
+/// "time": "2022-12-28T11:07:19.317310284Z",
+/// "idempotency_key": "2022-12-28 11:07:19.317310324 UTC-1-4019",
+/// "value": 12345454
+/// }
+/// ```
+
+/// ProxyConsumptionMetric struct that defines the format for one
+/// proxy metric entry
+/// i.e.
+///
+/// ```json
+/// {
+/// "metric": "remote_storage_size",
+/// "type": "incremental",
+/// "endpoint_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+/// "direction": "rx",
+/// "start_time": "2022-12-28T11:07:19.317310284Z",
+/// "stop_time": "2022-12-28T11:07:29.317310284Z",
+/// "idempotency_key": "2022-12-28 11:07:19.317310324 UTC-1-4019",
+/// "value": 12345454
+/// }
+/// ```
+
+pub fn idempotency_key(node_id: String) -> String {
+    format!(
+        "{}-{}-{:04}",
+        Utc::now(),
+        node_id,
+        rand::thread_rng().gen_range(0..=9999)
+    )
+}
+
+/// Key that uniquely identifies the object, this metric describes.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PageserverConsumptionMetricsKey {
+    pub tenant_id: TenantId,
+    pub timeline_id: Option<TimelineId>,
+    pub metric: &'static str,
+}
```

Comment:
What do you think about moving this to the pageserver? Same for `ProxyConsumptionMetricsKey`. That'd make this crate even smaller, but I'm reasonably confident that there is some logic that we could put here (maybe later).